node/src/components/linear_chain_sync.rs

```diff
@@ -301,127 +313,152 @@ where
                     State::SyncingTrustedHash { trusted_hash, .. } => {
                         trace!(?trusted_hash, "start synchronization");
                         // Start synchronization.
+                        self.metrics.reset_start_time();
                         fetch_block_by_hash(effect_builder, init_peer, trusted_hash)
                     }
                 }
             }
-            Event::GetBlockHeightResult(block_height, fetch_result) => match fetch_result {
-                BlockByHeightResult::Absent(peer) => {
-                    trace!(%block_height, %peer, "failed to download block by height. Trying next peer");
-                    self.peers.failure(&peer);
-                    match self.peers.random() {
-                        None => {
-                            // `block_height` not found on any of the peers.
-                            // We have synchronized all, currently existing, descendants of trusted
-                            // hash.
-                            self.mark_done();
-                            info!("finished synchronizing descendants of the trusted hash.");
-                            Effects::new()
+            Event::GetBlockHeightResult(block_height, fetch_result) => {
+                match fetch_result {
+                    BlockByHeightResult::Absent(peer) => {
```

Comment:
You could save two levels of indentation by merging the `match` statements (also below):
```rust
Event::GetBlockHeightResult(block_height, BlockByHeightResult::Absent(peer)) => {
```

But maybe it would make sense to split this whole method up and extract one method per `match` arm? (Not necessarily in this PR, of course!)

Reply:
Another PR please ;)